### ISSUE TITLE
Nuget fails to read local cache under macOS 10.12 (Sierra)

### ISSUE THREAD

**Issue by brendandixon** [NuGet/Home#5101]
### Overview
Using .NET Core 1.1.1 (installed via package dotnet-dev-osx-x64.1.0.1.pkg), Nuget (via dotnet) fails to "see" the local cache. The problem makes itself apparent in Omnisharp which fails to see restored assemblies unless restored to the global cache.

### Detailed Setup
I use a Docker container to build / test code that I develop under macOS using VSCode. To enable both sides to "see" the assemblies, I have a `NuGet.config` file specifying to restore assemblies to a project-local directory (i.e., .nuget). This worked quite well prior to .NET Core 1.1.1 (e.g., .NET Core 1.03/4 worked fine).

With .NET Core 1.1.1, the Linux container correctly sees the restored assemblies (and restores to the correct location). On macOS it does not (even thought it claims to use the local `NuGet.config` file).

Here is my `NuGet.config`:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <config>
    <add key="repositoryPath" value="./.nuget" />
  </config>
</configuration>
```

I also tried replacing `repositoryPath` with `globalPackagesFolder`; it made no difference.

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): dotnet

NuGet version (x.x.x.xxx): n/a

dotnet.exe --version (if appropriate): 1.0.1

`dotnet --info` for Linux:
```
.NET Command Line Tools (1.0.1)
 
Product Information:
Version:            1.0.1
Commit SHA-1 hash:  005db40cd1
 
Runtime Environment:
OS Name:     debian
OS Version:  8
OS Platform: Linux
RID:         debian.8-x64
Base Path:   /usr/share/dotnet/sdk/1.0.1
```

`dotnet --info` for macOS:
```
.NET Command Line Tools (1.0.1)
 
Product Information:
Version:            1.0.1
Commit SHA-1 hash:  005db40cd1
 
Runtime Environment:
OS Name:     Mac OS X
OS Version:  10.12
OS Platform: Darwin
RID:         osx.10.12-x64
Base Path:   /usr/local/share/dotnet/sdk/1.0.1
```

VS version (if appropriate): VSCode 1.11.2

Omnisharp version: 1.8.1

OS version (i.e. win10 v1607 (14393.321)): see `dotnet --info` output above

Worked before? If so, with which NuGet version: .NET Core 1.0.3 / 1.0.4

## Detailed repro steps so we can see the same problem
1. Create a new .NET Core project on macOS (e.g., `dotnet new classlib`)
2. Delete `~/.nuget` and run `dotnet version` (this clears the entire NuGet cache and recreates the basics .NET Core requires)
3. Add a `NuGet.config` using the Xml above
4. Run `dotnet restore` in the project
5. Note that the restore puts the assemblies in `~/.nuget` and not the project-local folder

**Comment by emgarten** [comment]
Use ``globalPackagesFolder`` to set the package folder path, dotnet does not use ``repositoryPath`` since it is only for packages.config projects.

Does restore list the NuGet.Config file you modified as a contributing config? My guess is that you may have this file in the wrong location and it isn't being used.

**Comment by brendandixon** [comment]
@emgarten The detail I included above both notes that I tried using `globalPackagesFolder`. The output from `dotnet nuget locals --list` also shows both Linux and macOS *claiming* to abide by `NuGet.config`. But on macOS it has no effect.

Did you attempt the repro steps I outlined?

**Comment by zhili1208** [comment]
@brendandixon "dotnet nuget lcoals --list" always show default location(~/.nuget), I think this is a bug.
I tried the repro steps, it's working on mac by using "globalPackagesFolder" .
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <config>
    <add key="globalPackagesFolder" value=".nuget" />
  </config>
</configuration>
```
dotnet restore puts all packages in ./.nuget/. 
dotnet version is 2.0.0-preview1-005791
OS version is Mac OS X 10.12

**Comment by zhili1208** [comment]
@mishra14 
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/LocalsCommand.cs#L45
**BUG**: "dotnet nuget lcoals --list" always show default location(~/.nuget)

**Comment by brendandixon** [comment]
@zhili1208 The problem occurs under .NET Core 1.1.1 (as noted above). You appear to have tested under .NET Core 2.x.

**Comment by brendandixon** [comment]
@mishra14 @zhili1208 I think I better understand the issues. As noted above, I use containers to build my code. *This is an increasingly common development practice.* Sadly, the restore process writes *absolute* paths into the `project.assets.json` file. Those paths are good only in the context of the container and not the host environment. Editing the code (say, using Omnisharp in VSCode) fails to see the installed packages because the paths in `project.assets.json` are not present.

This, to me, seems a significant failure of the how projects work. All paths within a project, especially those referencing locally installed packages and built artifacts, *must* be project-relative and not absolute. Using absolute paths is, generally, a bad idea.

Here is a `diff` of `project.assets.json` showing how the paths change based on the restore. Note that *none of these paths needs to be absolute as all are relative to the project*:

```
[malbec:AzureAgent (linux *=)] $ diff Microsoft.Azure.Agent.Common/obj/project.assets.json ~/Desktop/project.assets.json
38413c38413
<     "/Users/brendan/Code/AzureAgent/.nuget": {}
---
>     "/root/agent/.nuget": {}
38418c38418
<       "projectUniqueName": "/Users/brendan/Code/AzureAgent/Microsoft.Azure.Agent.Common/Microsoft.Azure.Agent.Common.csproj",
---
>       "projectUniqueName": "/root/agent/Microsoft.Azure.Agent.Common/Microsoft.Azure.Agent.Common.csproj",
38420,38421c38420,38421
<       "projectPath": "/Users/brendan/Code/AzureAgent/Microsoft.Azure.Agent.Common/Microsoft.Azure.Agent.Common.csproj",
<       "outputPath": "/Users/brendan/Code/AzureAgent/Microsoft.Azure.Agent.Common/obj/",
---
>       "projectPath": "/root/agent/Microsoft.Azure.Agent.Common/Microsoft.Azure.Agent.Common.csproj",
>       "outputPath": "/root/agent/Microsoft.Azure.Agent.Common/obj/",
38427c38427
<         "lib/netstandard1.6/Microsoft.Azure.Agent.Common.dll": "/Users/brendan/Code/AzureAgent/Microsoft.Azure.Agent.Common/bin/placeholder/netstandard1.6/Microsoft.Azure.Agent.Common.dll"
---
>         "lib/netstandard1.6/Microsoft.Azure.Agent.Common.dll": "/root/agent/Microsoft.Azure.Agent.Common/bin/placeholder/netstandard1.6/Microsoft.Azure.Agent.Common.dll"
```

**Comment by mishra14** [comment]
@emgarten : Is this a problem in restore?

**Comment by emgarten** [comment]
@mishra14 @brendandixon absolute paths in project.assets.json for docker is covered by this issue: https://github.com/NuGet/Home/issues/4872

Let's keep this issue for the nuget locals change

**Comment by rrelyea** [comment]
@mishra14  - please confirm, I believe this fix shipped in 4.3

**Comment by mishra14** [comment]
This was merged on April 26. So this should be in 4.3.
